src/wh_server_crypto.c

```diff
@@ -841,6 +848,57 @@ static int hsmCryptoCmac(whServerContext* server, whPacket* packet,
 }
 #endif
 
+#ifndef NO_SHA256
+static int hsmCryptoSha256(whServerContext* server, whPacket* packet,
+                           uint16_t* size)
+{
+    int                        ret    = 0;
+    wc_Sha256*                 sha256 = server->crypto->algoCtx.sha256;
+    wh_Packet_hash_sha256_req* req    = &packet->hashSha256Req;
+    wh_Packet_hash_sha256_res* res    = &packet->hashSha256Res;
+
+    /* THe server SHA256 struct doesn't persist state (it is a union), meaning
+     * the devId may get blown away between calls. We must restore the server
+     * devId each time */
+    sha256->devId = server->crypto->devId;
+
+    /* Init the SHA256 context if this is the first time, otherwise restore the
+     * hash state from the client */
+    if (req->resumeState.resumeHashFlag == 0) {
```

Comment:
should this be a bitfield like RESUME_HASH_BIT_FLAG?  Not sure we really need anything else in a flag for SHA2

Wouldn't the server already know if this is a resumption based on the hi_len/lo_len > BLOCK_SIZE?

Reply:
Ah okay, I remember why I did it this way - the empty hash. Sha256 hash of 0 length input data does result in a valid digest and my thoughts were it would need to be held by the client too. But I forgot that a zero length message wouldn't be even sent across the wire since we only send full blocks, and the way sha256 works is that the zero length message, though it produces a valid hash on finalization, doesn't affect the internal state of the hash at all. So it wasn't needed like you said.